inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/PulsarSource.java

```diff
@@ -0,0 +1,312 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.apache.inlong.agent.constant.CommonConstants.DEFAULT_PROXY_PACKAGE_MAX_SIZE;
+import static org.apache.inlong.agent.constant.CommonConstants.PROXY_KEY_DATA;
+import static org.apache.inlong.agent.constant.CommonConstants.PROXY_KEY_STREAM_ID;
+import static org.apache.inlong.agent.constant.CommonConstants.PROXY_PACKAGE_MAX_SIZE;
+import static org.apache.inlong.agent.constant.CommonConstants.PROXY_SEND_PARTITION_KEY;
+import static org.apache.inlong.agent.constant.FetcherConstants.AGENT_GLOBAL_READER_QUEUE_PERMIT;
+import static org.apache.inlong.agent.constant.FetcherConstants.AGENT_GLOBAL_READER_SOURCE_PERMIT;
+import static org.apache.inlong.agent.constant.TaskConstants.OFFSET;
+import static org.apache.inlong.agent.constant.TaskConstants.TASK_CYCLE_UNIT;
+import static org.apache.inlong.agent.constant.TaskConstants.TASK_PULSAR_RESET_TIME;
+import static org.apache.inlong.agent.constant.TaskConstants.TASK_PULSAR_SCAN_STARTUP_MODE;
+import static org.apache.inlong.agent.constant.TaskConstants.TASK_PULSAR_SERVICE_URL;
+
+public class PulsarSource extends AbstractSource {
+
+    @Data
+    @AllArgsConstructor
+    @NoArgsConstructor
+    private class SourceData {
+
+        private byte[] data;
+        private Long offset;
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PulsarSource.class);
+    private static final ThreadPoolExecutor EXECUTOR_SERVICE = new ThreadPoolExecutor(
+            0, Integer.MAX_VALUE,
+            1L, TimeUnit.SECONDS,
+            new SynchronousQueue<>(),
+            new AgentThreadFactory("pulsar-source"));
+    private BlockingQueue<SourceData> queue;
+    public InstanceProfile profile;
+    private int maxPackSize;
+    private String taskId;
+    private String instanceId;
+    private String topic;
+    private String serviceUrl;
+    private String scanStartupMode;
+    private PulsarClient pulsarClient;
+    private Long timestamp;
+    Map<Integer, Long> partitionOffsets = new HashMap<>();
+    private volatile boolean running = false;
+    private volatile boolean runnable = true;
+    private volatile AtomicLong emptyCount = new AtomicLong(0);
+
+    private final Integer CACHE_QUEUE_SIZE = 100000;
+    private final Integer READ_WAIT_TIMEOUT_MS = 10;
+    private final Integer EMPTY_CHECK_COUNT_AT_LEAST = 5 * 60;
+    private final Integer BATCH_TOTAL_LEN = 1024 * 1024;
+
+    private final Integer CORE_THREAD_PRINT_INTERVAL_MS = 1000;
+    private boolean isRealTime = false;
+    private boolean isRestoreFromDB = false;
```

Comment:
`isRestoreFromDB` is always false, is it ok?

Reply:
I have fixed it.